substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrThreadLocal.java

```diff
@@ -208,10 +208,21 @@ public JfrBuffer getNativeBuffer() {
     @Uninterruptible(reason = "Accesses a JFR buffer.")
     public static JfrBuffer flush(JfrBuffer threadLocalBuffer, UnsignedWord uncommitted, int requested) {
         assert threadLocalBuffer.isNonNull();
+        // TODO using acquired field is kind of a hack of the field knowing that
+        // it won't be used for thread local buffers elsewhere (yet?). Ideally
+        // there is similar synchronization in JfrBufferAccess to acquire top
+        boolean acquired = JfrBufferAccess.acquire(threadLocalBuffer);
+        while (!acquired) {
```

Comment:
Hm. I'm confused here since after this adjustment, both thread A and B are trying to synchronize now *only* in uninterruptible code blocks. Or at least that was the intention... If it could block this way then how should a critical section be protected for uninterruptible code?

Reply:
The mutex is now only used in uninterruptible code but the new code still has a very similar issue as before. This is the case because you replaced the mutex with a different concept that still blocks (the blocking is only done in a different way). Let me try to be more specific:
- Thread A executes `JfrThreadLocalMemory.writeBuffers()` and acquires buffer X of thread B.
- Thread B wants to flush buffer X. For that, it executes the `while` loop in `JfrThreadLocal.flush()`, which is in uninterruptible code and does busy waiting until it can acquire buffer X.
- Thread A still holds buffer X but needs a safepoint. So, a safepoint gets scheduled and all threads block that are outside of uninterruptible code.
- Thread B is still stuck in uninterruptible code. So, the VM can never reach a safepoint and deadlocks.

Reply:
Oh, I'm dumb. Right my intention was to only synchronize in uninterruptible code, but the acquire/release in writeBuffers is not. Ugh my bad; for some reason I thought I had made that uninterruptible too. Back to the drawing board I go, sorry about this!

Reply:
I guess this goes on top of reading/writing files uninterruptibly then...

Reply:
@christianhaeubl Thanks Christian for your time and reviews here! I'll continue this after the initial merge.